okhttp/src/test/java/io/grpc/okhttp/OkHttpChannelBuilderTest.java

```diff
@@ -155,6 +165,110 @@ public void sslSocketFactoryFrom_unsupportedTls() {
     assertThat(result.factory).isNull();
   }
 
+  @Test
+  public void sslSocketFactoryFrom_tls_customRoots() throws Exception {
+    SelfSignedCertificate cert = new SelfSignedCertificate(TestUtils.TEST_SERVER_HOST);
+    KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+    keyStore.load(null);
+    keyStore.setKeyEntry("mykey", cert.key(), new char[0], new Certificate[] {cert.cert()});
+    KeyManagerFactory keyManagerFactory =
+        KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+    keyManagerFactory.init(keyStore, new char[0]);
+
+    SSLContext serverContext = SSLContext.getInstance("TLS");
+    serverContext.init(keyManagerFactory.getKeyManagers(), null, null);
+    final SSLServerSocket serverListenSocket =
+        (SSLServerSocket) serverContext.getServerSocketFactory().createServerSocket(0);
+    final SettableFuture<SSLSocket> serverSocket = SettableFuture.create();
+    new Thread(new Runnable() {
+      @Override public void run() {
+        try {
+          SSLSocket socket = (SSLSocket) serverListenSocket.accept();
+          socket.getSession(); // Force handshake
+          serverSocket.set(socket);
+          serverListenSocket.close();
+        } catch (Throwable t) {
+          serverSocket.setException(t);
+        }
+      }
+    }).start();
+
+    ChannelCredentials creds = TlsChannelCredentials.newBuilder()
```

Comment:
How about a negative test also: client doesn't trust the server cert i.e. trustManager not initialized with `cert`